### PR TITLE
many: use the new install/refresh API by switching snapstate to use store.SnapAction

### PR DESCRIPTION
**Pull request by pedronis**
this could be split in two PR if preferred:  switching refreshes to the new API and then install-like operations.

### CODE REVIEW

**Comment by pedronis** [comment]
based on both #4896 and #4772

File: store/errors.go [hunk 1]
```diff
@@ -107,3 +110,67 @@ func (e InvalidAuthDataError) Error() string {
 	//      (empirically this checks out)
 	return strings.Join(es, "  ")
 }
+
+// InstallRefreshError conveys errors that were reported on otherwise overall successful install/refresh request.
+type InstallRefreshError struct {
```
Comment (niemeyer) [inline store/errors.go]: The term "InstallRefresh" feels a little bit awkward, in the sense that this is a pair of operations, yet it reads like an action ("install the refresh"). This PR is using it in quite a few places, and I think we're adding it just now mostly, so perhaps the right time to rethink. Can we come up with a more general term that would convey either operation?  Perhaps "SnapAction", which leverages the term already being adopted as part of these? So we'd have SnapActionError, store.SnapAction, Store.SnapAction(...), etc.

File: tests/lib/fakestore/store/store.go [hunk 2]
```diff
@@ -476,6 +478,153 @@ func (s *Store) collectAssertions() (asserts.Backstore, error) {
 	return bs, nil
 }
 
+type contextSnap struct {
```
Comment (niemeyer) [inline tests/lib/fakestore/store/store.go]: currentSnap, assuming we go with the proposal above.

File: tests/lib/fakestore/store/store.go [hunk 3]
```diff
@@ -476,6 +478,153 @@ func (s *Store) collectAssertions() (asserts.Backstore, error) {
...
+}
+
+type installRefreshAction struct {
+	Action      string `json:"action"`
+	InstanceKey string `json:"instance-key"`
+	SnapID      string `json:"snap-id"`
+	Name        string `json:"name"`
+}
+
+type installRefreshReqJSON struct {
```
Comment (niemeyer) [inline tests/lib/fakestore/store/store.go]: Everything here is "JSON".. is that a Request to the store while the one below is the Result? If so just Request might do. Similar comments for the types below.

File: tests/lib/fakestore/store/store.go [hunk 4]
```diff
@@ -476,6 +478,153 @@ func (s *Store) collectAssertions() (asserts.Backstore, error) {
...
+
+type installRefreshResult struct {
+	Result      string        `json:"result"`
+	InstanceKey string        `json:"instance-key"`
+	SnapID      string        `json:"snap-id"`
+	Name        string        `json:"name"`
+	Snap        detailsV2JSON `json:"snap"`
+}
+
+type detailsV2JSON struct {
```
Comment (niemeyer) [inline tests/lib/fakestore/store/store.go]: detailsResultV2, or something?

File: tests/lib/fakestore/store/store.go [hunk 5]
```diff
@@ -476,6 +478,153 @@ func (s *Store) collectAssertions() (asserts.Backstore, error) {
...
+	Download struct {
+		URL      string `json:"url"`
+		Sha3_384 string `json:"sha3-384"`
+		Size     uint64 `json:"size"`
+	} `json:"download"`
+	Version  string `json:"version"`
+	Revision int    `json:"revision"`
+}
+
+type installRefreshRespJSON struct {
```
Comment (niemeyer) [inline tests/lib/fakestore/store/store.go]: snapActionResultList?  Should probably be next to snapActionResult above.

File: store/errors.go (hunk 1 above)
Comment (pedronis) [reply store/errors.go]: used SnapAction

File: overlord/snapstate/handlers.go [hunk 6]
```diff
@@ -362,6 +361,12 @@ func (m *SnapManager) undoPrepareSnap(t *state.Task, _ *tomb.Tomb) error {
 	return nil
 }
 
+func installInfoUnlocked(st *state.State, snapsup *SnapSetup) (*snap.Info, error) {
+	st.Lock()
```
[the remark this answers is not in the document]
Comment (pedronis) [reply overlord/snapstate/handlers.go]: is the same meaning  of Unlocked in NewTaskProgressAdapterUnlocked , I followed that.

I used the small helper because on panics (which are not unlikely in tests given how some helper are written) wrapping the call tends to produce a bit less readable stacktraces that first panic about the lock state

File: cmd/snap/cmd_snap_op.go [hunk 7]
```diff
@@ -284,7 +284,7 @@ func showDone(names []string, op string) error {
 		default:
 			fmt.Fprintf(Stdout, "internal error: unknown op %q", op)
 		}
-		if snap.TrackingChannel != snap.Channel {
+		if snap.TrackingChannel != snap.Channel && snap.Channel != "" {
```
Comment (chipaca) [inline cmd/snap/cmd_snap_op.go]: I understand this change, but I don't understand its presence on this PR. Drive-by? Bad merge? (logic seems fine though)

File: daemon/api.go [hunk 8]
```diff
@@ -1165,7 +1165,13 @@ func (inst *snapInstruction) errToResponse(err error) Response {
...
 
 	switch err {
-	case store.ErrSnapNotFound:
+	case store.ErrSnapNotFound, store.ErrRevisionNotAvailable:
+		// TODO: treating ErrRevisionNotAvailable the same as
+		// ErrSnapNotFound preserves the old error handling
+		// behavior of the REST API and the snap command.  We
+		// should revisit this once the store returns more
+		// precise errors and makes it possible to distinguish
+		// the why a revision wasn't available.
```
Comment (chipaca) [inline daemon/api.go]: ♥

File: tests/lib/fakestore/store/store.go [hunk 9]
```diff
@@ -185,19 +187,19 @@ func snapEssentialInfo(w http.ResponseWriter, fn, snapID string, bs asserts.Back
 
 	info, err := snap.ReadInfoFromSnapFile(snapFile, nil)
 	if err != nil {
-		http.Error(w, fmt.Sprintf("can get info for: %v: %v", fn, err), 400)
+		http.Error(w, fmt.Sprintf("cannot get info for: %v: %v", fn, err), 400)
```
Comment (chipaca) [inline tests/lib/fakestore/store/store.go]: heh

File: cmd/snap/cmd_snap_op.go (hunk 7 above)
Comment (pedronis) [reply cmd/snap/cmd_snap_op.go]: the new api will return an empty channel when doing install by revision, the old API was just returning a bit of a non-sense guess. Some part of this code were already checking for snap:Channel empty but not this one.

File: overlord/snapstate/snapmgr.go [hunk 10]
```diff
@@ -230,13 +231,28 @@ func readInfoAnyway(name string, si *snap.SideInfo) (*snap.Info, error) {
...
 // CurrentInfo returns the information about the current active revision or the last active revision (if the snap is inactive). It returns the ErrNoCurrent error if snapst.Current is unset.
 func (snapst *SnapState) CurrentInfo() (*snap.Info, error) {
 	cur := snapst.CurrentSideInfo()
 	if cur == nil {
 		return nil, ErrNoCurrent
 	}
-	return readInfo(cur.RealName, cur)
+	info, err := readInfo(cur.RealName, cur)
+	if err != nil {
+		logger.Noticef("%v", err)
```
[the remark this answers is not in the document]
Comment (pedronis) [reply overlord/snapstate/snapmgr.go]: that was from a comment from gustavo

File: overlord/snapstate/storehelpers.go [hunk 11]
```diff
@@ -90,56 +90,82 @@ func userFromUserIDOrFallback(st *state.State, userID int, fallbackUser *auth.Us
...
 	}
-	theStore := Store(st)
-	st.Unlock() // calls to the store should be done without holding the state lock
-	spec := store.SnapSpec{
-		Name:     name,
-		Channel:  channel,
+
+	// cannot specify both with the API
+	if !revision.Unset() {
+		channel = ""
```
[the remark this answers is not in the document]
Comment (pedronis) [reply overlord/snapstate/storehelpers.go]: the user can do  --revision=3 --channel=stable,  but the server doesn't want both and errors, we need a bit to decide where to ignore the channel,  here seems the best place

File: overlord/hookstate/ctlcmd/services_test.go [hunk 12]
```diff
@@ -47,18 +47,18 @@ type fakeStore struct {
...
+			SideInfo: snap.SideInfo{
+				RealName: actions[0].Name,
+				Revision: snap.R(2),
+			},
+			Publisher:     "foo",
+			Architectures: []string{"all"},
+		}}, nil
+	}
 
-func (f *fakeStore) ListRefresh(_ context.Context, cand []*store.RefreshCandidate, user *auth.UserState, opt *store.RefreshOptions) ([]*snap.Info, error) {
```
[the remark this answers is not in the document]
Comment (pedronis) [reply overlord/hookstate/ctlcmd/services_test.go]: don't think so, no code in this package calls SnapAction directly, but through snapstate functions tested elsewhere

File: overlord/snapstate/backend_test.go [hunk 13]
```diff
@@ -240,52 +281,150 @@ func (f *fakeStore) LookupRefresh(cand *store.RefreshCandidate, user *auth.UserS
...
+	sort.Sort(sorted)
+
+	refreshErrors := make(map[string]error)
+	installErrors := make(map[string]error)
 	var res []*snap.Info
-	for _, cand := range cands {
-		info, err := f.LookupRefresh(cand, user)
-		if err == store.ErrLocalSnap || err == store.ErrNoUpdateAvailable {
+	for _, a := range sorted {
+		if a.Action == "install" {
```
[the remark this answers is not in the document]
Comment (pedronis) [reply overlord/snapstate/backend_test.go]: those helpers would need to take quite a bit of mutable input args, and the switch would grow indentation a bit. so not sure I have brilliant idea how to touch yet

File: overlord/snapstate/backend_test.go [hunk 14]
```diff
@@ -240,52 +281,150 @@ func (f *fakeStore) LookupRefresh(cand *store.RefreshCandidate, user *auth.UserS
...
+				revno:  info.Revision,
+				userID: userID,
+			})
+			if !a.Revision.Unset() {
+				info.Channel = ""
+			}
+			res = append(res, info)
+			continue
+		}
+		if a.Action != "refresh" {
```
[the remark this answers is not in the document]
Comment (pedronis) [reply overlord/snapstate/backend_test.go]: yes, I think I can move the check at the top

File: overlord/snapstate/autorefresh_test.go [hunk 15]
```diff
@@ -57,6 +57,14 @@ func (r *autoRefreshStore) ListRefresh(ctx context.Context, cands []*store.Refre
 	return nil, r.listRefreshErr
 }
 
+func (r *autoRefreshStore) SnapAction(ctx context.Context, currentSnaps []*store.CurrentSnap, actions []*store.SnapAction, user *auth.UserState, opts *store.RefreshOptions) ([]*snap.Info, error) {
+	if ctx == nil || !auth.IsEnsureContext(ctx) {
+		panic("Ensure marked context required")
+	}
+	r.ops = append(r.ops, "list-refresh")
```
[the remark this answers is not in the document]
Comment (pedronis) [reply overlord/snapstate/autorefresh_test.go]: the caller goes through AutoRefresh, I don't remember if that has its own tests though

File: overlord/snapstate/refreshhints_test.go [hunk 16]
```diff
@@ -49,6 +49,14 @@ func (r *recordingStore) ListRefresh(ctx context.Context, cands []*store.Refresh
 	return nil, nil
 }
 
+func (r *recordingStore) SnapAction(ctx context.Context, currentSnaps []*store.CurrentSnap, actions []*store.SnapAction, user *auth.UserState, opts *store.RefreshOptions) ([]*snap.Info, error) {
+	if ctx == nil || !auth.IsEnsureContext(ctx) {
+		panic("Ensure marked context required")
+	}
```
[the remark this answers is not in the document]
Comment (pedronis) [reply overlord/snapstate/refreshhints_test.go]: here probably yes, the code is calling a low level helper directly

File: overlord/snapstate/backend_test.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (pedronis) [reply overlord/snapstate/backend_test.go]: done

File: overlord/snapstate/refreshhints_test.go (hunk 16 above)
[the remark this answers is not in the document]
Comment (pedronis) [reply overlord/snapstate/refreshhints_test.go]: done

File: overlord/snapstate/autorefresh_test.go (hunk 15 above)
[the remark this answers is not in the document]
Comment (pedronis) [reply overlord/snapstate/autorefresh_test.go]: seems is not well tested on its own, done

File: store/errors.go [hunk 17]
```diff
@@ -182,9 +182,6 @@ var (
 func translateSnapActionError(action, code, message string) error {
 	switch code {
 	case "revision-not-found":
-		if action == "refresh" {
-			return ErrNoUpdateAvailable
```
Comment (mvo5) [inline store/errors.go]: I assume this is just not needed anymore, right? I.e. we still return ErrNoUpdateAvailable at the right place.

File: tests/lib/fakestore/store/store.go [hunk 18]
```diff
@@ -482,6 +484,153 @@ func (s *Store) collectAssertions() (asserts.Backstore, error) {
...
+			res.Snap.Publisher.Username = essInfo.DevelName
+			res.Snap.Download.URL = fmt.Sprintf("%s/download/%s", s.URL(), filepath.Base(fn))
+			res.Snap.Download.Sha3_384 = hexify(essInfo.Digest)
+			res.Snap.Download.Size = essInfo.Size
+			replyData.Results = append(replyData.Results, res)
+		}
+	}
+
+	// use indent because this is a development tool, output
+	// should look nice
```
Comment (mvo5) [inline tests/lib/fakestore/store/store.go]: :heart:

File: overlord/snapstate/storehelpers.go [hunk 19]
```diff
@@ -157,39 +183,127 @@ func preUpdateInfo(st *state.State, snapst *SnapState, amend bool, userID int) (
...
-		// in amend mode we need to move to the store rev
-		id, err := snapNameToID(st, curInfo.Name(), user)
-		if err != nil {
-			return nil, nil, fmt.Errorf("cannot get snap ID for %q: %v", curInfo.Name(), err)
+	return curInfo, user, nil
+}
+
+func singleActionResult(name, action string, results []*snap.Info, e error) (info *snap.Info, err error) {
+	if len(results) > 0 {
+		// TODO: if we also have an error log/warn about it
```
[the remark this answers is not in the document]
Comment (pedronis) [reply overlord/snapstate/storehelpers.go]: best to have an error, probably before merging:

internal error: multiple store result in the context of a single snap op

or something

File: store/errors.go (hunk 17 above)
Comment (pedronis) [reply store/errors.go]: correct, I moved the backward compat logic all in daemon

**Comment by pedronis** [comment]
need to add a small check and do a master merge before squash merging
